File: Sources/XcodeSimulatorLegacyHost/MenuController.m
#import "MenuController.h"

#import "DeviceWindowController.h"

static NSString *XSHFunctionKeyEquivalent(unichar character) {
    return [NSString stringWithCharacters:&character length:1];
}

@implementation XSHMenuController

- (void)installMainMenu {
    NSString *applicationName = @"Xcode Simulator Legacy Host";
    NSMenu *mainMenu = [[NSMenu alloc] initWithTitle:@""];

    NSMenu *applicationMenu = [self addMenuWithTitle:applicationName
                                         toMainMenu:mainMenu];
    [self addItemWithTitle:[@"About " stringByAppendingString:applicationName]
                    action:@selector(orderFrontStandardAboutPanel:)
             keyEquivalent:@""
                 modifiers:0
                     toMenu:applicationMenu];
    [applicationMenu addItem:NSMenuItem.separatorItem];
    [self addItemWithTitle:[@"Hide " stringByAppendingString:applicationName]
                    action:@selector(hide:)
             keyEquivalent:@"h"
                 modifiers:NSEventModifierFlagCommand
                     toMenu:applicationMenu];
    [self addItemWithTitle:@"Hide Others"
                    action:@selector(hideOtherApplications:)
             keyEquivalent:@"h"
                 modifiers:NSEventModifierFlagOption | NSEventModifierFlagCommand
                     toMenu:applicationMenu];
    [self addItemWithTitle:@"Show All"
                    action:@selector(unhideAllApplications:)
             keyEquivalent:@""
                 modifiers:0
                     toMenu:applicationMenu];
    [applicationMenu addItem:NSMenuItem.separatorItem];
    [self addItemWithTitle:[@"Quit " stringByAppendingString:applicationName]
                    action:@selector(quit:)
             keyEquivalent:@"q"
                 modifiers:NSEventModifierFlagCommand
                     toMenu:applicationMenu];

    NSMenu *fileMenu = [self addMenuWithTitle:@"File" toMainMenu:mainMenu];
    [self addItemWithTitle:@"Save Screen…"
                    action:@selector(saveFramebufferImage:)
             keyEquivalent:@"s"
                 modifiers:NSEventModifierFlagCommand
                     toMenu:fileMenu];
    [fileMenu addItem:NSMenuItem.separatorItem];
    [self addItemWithTitle:@"Close Window"
                    action:@selector(performClose:)
             keyEquivalent:@"w"
                 modifiers:NSEventModifierFlagCommand
                     toMenu:fileMenu];

    NSMenu *deviceMenu = [self addMenuWithTitle:@"Device" toMainMenu:mainMenu];
    [self addItemWithTitle:@"Rotate Left"
                    action:@selector(toggleRotationLeft:)
             keyEquivalent:XSHFunctionKeyEquivalent(NSLeftArrowFunctionKey)
                 modifiers:NSEventModifierFlagCommand
                     toMenu:deviceMenu];
    [self addItemWithTitle:@"Rotate Right"
                    action:@selector(toggleRotationRight:)
             keyEquivalent:XSHFunctionKeyEquivalent(NSRightArrowFunctionKey)
                 modifiers:NSEventModifierFlagCommand
                     toMenu:deviceMenu];
    [deviceMenu addItem:NSMenuItem.separatorItem];
    [self addItemWithTitle:@"Home"
                    action:@selector(homeButtonPressed:)
             keyEquivalent:@"h"
                 modifiers:NSEventModifierFlagShift | NSEventModifierFlagCommand
                     toMenu:deviceMenu];
    [self addItemWithTitle:@"Lock"
                    action:@selector(lockButtonPressed:)
             keyEquivalent:@"l"
                 modifiers:NSEventModifierFlagCommand
                     toMenu:deviceMenu];
    [self addItemWithTitle:@"Shake"
                    action:@selector(shakeDevice:)
             keyEquivalent:@"z"
                 modifiers:NSEventModifierFlagControl | NSEventModifierFlagCommand
                     toMenu:deviceMenu];

    NSMenu *inputOutputMenu = [self addMenuWithTitle:@"I/O" toMainMenu:mainMenu];
    [self addItemWithTitle:@"Toggle Software Keyboard"
                    action:@selector(toggleSoftwareKeyboard:)
             keyEquivalent:@"k"
                 modifiers:NSEventModifierFlagCommand
                     toMenu:inputOutputMenu];

    NSMenu *featuresMenu = [self addMenuWithTitle:@"Features" toMainMenu:mainMenu];
    [self addItemWithTitle:@"Toggle Appearance"
                    action:@selector(toggleAppearance:)
             keyEquivalent:@"a"
                 modifiers:NSEventModifierFlagShift | NSEventModifierFlagCommand
                     toMenu:featuresMenu];

    NSMenu *windowMenu = [self addMenuWithTitle:@"Window" toMainMenu:mainMenu];
    [self addItemWithTitle:@"Minimize"
                    action:@selector(performMiniaturize:)
             keyEquivalent:@"m"
                 modifiers:NSEventModifierFlagCommand
                     toMenu:windowMenu];
    [self addItemWithTitle:@"Zoom"
                    action:@selector(performZoom:)
             keyEquivalent:@""
                 modifiers:0
                     toMenu:windowMenu];
    [self addItemWithTitle:@"Enter Full Screen"
                    action:@selector(toggleFullScreen:)
             keyEquivalent:@"f"
                 modifiers:NSEventModifierFlagControl | NSEventModifierFlagCommand
                     toMenu:windowMenu];
    [self addItemWithTitle:@"Show Device Bezels"
                    action:@selector(toggleShowChrome:)
             keyEquivalent:@""
                 modifiers:0
                     toMenu:windowMenu];
    [self addItemWithTitle:@"Stay On Top"
                    action:@selector(toggleAlwaysOnTop:)
             keyEquivalent:@""
                 modifiers:0
                     toMenu:windowMenu];
    [windowMenu addItem:NSMenuItem.separatorItem];
    NSMenuItem *fitScreenItem = [self addItemWithTitle:@"Fit Screen"
                                                action:@selector(windowScaleChanged:)
                                         keyEquivalent:@"4"
                                             modifiers:NSEventModifierFlagCommand
                                                 toMenu:windowMenu];
    fitScreenItem.tag = 4111;
    [windowMenu addItem:NSMenuItem.separatorItem];
    [self addItemWithTitle:@"Bring All to Front"
                    action:@selector(arrangeInFront:)
             keyEquivalent:@""
                 modifiers:0
                     toMenu:windowMenu];

    NSApp.mainMenu = mainMenu;
    NSApp.windowsMenu = windowMenu;
}

- (NSMenu *)addMenuWithTitle:(NSString *)title toMainMenu:(NSMenu *)mainMenu {
    NSMenuItem *item = [[NSMenuItem alloc] initWithTitle:title
                                                  action:nil
                                           keyEquivalent:@""];
    NSMenu *menu = [[NSMenu alloc] initWithTitle:title];
    item.submenu = menu;
    [mainMenu addItem:item];
    return menu;
}

- (NSMenuItem *)addItemWithTitle:(NSString *)title
                          action:(SEL)action
                   keyEquivalent:(NSString *)keyEquivalent
                       modifiers:(NSEventModifierFlags)modifiers
                           toMenu:(NSMenu *)menu {
    NSMenuItem *item = [[NSMenuItem alloc] initWithTitle:title
                                                  action:action
                                           keyEquivalent:keyEquivalent];
    item.target = self;
    item.keyEquivalentModifierMask = modifiers;
    [menu addItem:item];
    return item;
}

- (nullable XSHDeviceWindowController *)activeSession {
    NSWindowController *controller = NSApp.keyWindow.windowController;
    if (![controller isKindOfClass:XSHDeviceWindowController.class]) {
        return nil;
    }
    return (XSHDeviceWindowController *)controller;
}

- (BOOL)validateMenuItem:(NSMenuItem *)menuItem {
    SEL action = menuItem.action;
    XSHDeviceWindowController *session = self.activeSession;

    if (action == @selector(orderFrontStandardAboutPanel:) ||
        action == @selector(hide:) ||
        action == @selector(hideOtherApplications:) ||
        action == @selector(unhideAllApplications:) ||
        action == @selector(quit:)) {
        return YES;
    }

    if (action == @selector(arrangeInFront:)) {
        return NSApp.windows.count > 0;
    }
    if (session == nil) {
        menuItem.state = NSControlStateValueOff;
        return NO;
    }

    if (action == @selector(toggleShowChrome:)) {
        menuItem.state = session.deviceChromeVisible
            ? NSControlStateValueOn
            : NSControlStateValueOff;
        return session.canPerformCommands;
    }
    if (action == @selector(toggleAlwaysOnTop:)) {
        menuItem.state = session.stayingOnTop
            ? NSControlStateValueOn
            : NSControlStateValueOff;
        return session.canPerformCommands;
    }
    if (action == @selector(saveFramebufferImage:) ||
        action == @selector(toggleRotationLeft:) ||
        action == @selector(toggleRotationRight:)) {
        return session.canPerformToolOperation;
    }
    if (action == @selector(windowScaleChanged:)) {
        return session.canPerformCommands &&
            !session.window.inLiveResize &&
            !(session.window.styleMask & NSWindowStyleMaskFullScreen);
    }
    if (action == @selector(toggleFullScreen:)) {
        BOOL isFullScreen = session.window.styleMask & NSWindowStyleMaskFullScreen;
        menuItem.title = isFullScreen ? @"Exit Full Screen" : @"Enter Full Screen";
        return session.window != nil;
    }
    if (action == @selector(performClose:) ||
        action == @selector(performMiniaturize:) ||
        action == @selector(performZoom:)) {
        return session.window != nil;
    }
    return session.canPerformCommands;
}

- (void)orderFrontStandardAboutPanel:(id)sender {
    [NSApp orderFrontStandardAboutPanel:sender];
}

- (void)hide:(id)sender {
    [NSApp hide:sender];
}

- (void)hideOtherApplications:(id)sender {
    [NSApp hideOtherApplications:sender];
}

- (void)unhideAllApplications:(id)sender {
    [NSApp unhideAllApplications:sender];
}

- (void)quit:(id)sender {
    [NSApp terminate:sender];
}

- (void)saveFramebufferImage:(id)sender {
    [self.activeSession saveFramebufferImage:sender];
}

- (void)performClose:(id)sender {
    [self.activeSession.window performClose:sender];
}

- (void)toggleRotationLeft:(id)sender {
    [self.activeSession toggleRotationLeft:sender];
}

- (void)toggleRotationRight:(id)sender {
    [self.activeSession toggleRotationRight:sender];
}

- (void)homeButtonPressed:(id)sender {
    [self.activeSession homeButtonPressed:sender];
}

- (void)lockButtonPressed:(id)sender {
    [self.activeSession lockButtonPressed:sender];
}

- (void)shakeDevice:(id)sender {
    [self.activeSession shakeDevice:sender];
}

- (void)toggleSoftwareKeyboard:(id)sender {
    [self.activeSession toggleSoftwareKeyboard:sender];
}

- (void)toggleAppearance:(id)sender {
    [self.activeSession toggleAppearance:sender];
}

- (void)performMiniaturize:(id)sender {
    [self.activeSession.window performMiniaturize:sender];
}

- (void)performZoom:(id)sender {
    [self.activeSession.window performZoom:sender];
}

- (void)toggleFullScreen:(id)sender {
    [self.activeSession.window toggleFullScreen:sender];
}

- (void)toggleShowChrome:(id)sender {
    [self.activeSession toggleShowChrome:sender];
}

- (void)toggleAlwaysOnTop:(id)sender {
    [self.activeSession toggleAlwaysOnTop:sender];
}

- (void)windowScaleChanged:(id)sender {
    [self.activeSession fitScreen:sender];
}

- (void)arrangeInFront:(id)sender {
    [NSApp arrangeInFront:sender];
}

@end

File: Sources/NeoSimulator/MenuController.swift
import AppKit

@MainActor
final class MenuController: NSObject, NSMenuItemValidation {
    private weak var application: HostApplication?
    private var activeSession: DeviceWindowController? { NSApp.keyWindow?.windowController as? DeviceWindowController }

    init(application: HostApplication) { self.application = application }

    func install() {
        let main = NSMenu()
        let app = menu("NeoSimulator", in: main)
        item("About NeoSimulator", action: #selector(NSApplication.orderFrontStandardAboutPanel(_:)), in: app, target: NSApp)
        app.addItem(.separator())
        item("Hide NeoSimulator", action: #selector(NSApplication.hide(_:)), key: "h", in: app, target: NSApp)
        item("Hide Others", action: #selector(NSApplication.hideOtherApplications(_:)), key: "h", modifiers: [.command, .option], in: app, target: NSApp)
        item("Show All", action: #selector(NSApplication.unhideAllApplications(_:)), in: app, target: NSApp)
        app.addItem(.separator())
        item("Quit NeoSimulator", action: #selector(NSApplication.terminate(_:)), key: "q", in: app, target: NSApp)

        let file = menu("File", in: main)
        item("Open Simulator…", action: #selector(showBrowser(_:)), key: "n", in: file)
        file.addItem(.separator())
        command("Save Screen…", .screenshot, key: "s", in: file)
        command("Record Video…", .recording, key: "r", in: file)
        file.addItem(.separator())
        command("Install App or Import Media…", .importFiles, key: "i", modifiers: [.command, .shift], in: file)
        file.addItem(.separator())
        item("Close Window", action: #selector(closeWindow(_:)), key: "w", in: file)

        let device = menu("Device", in: main)
        command("Rotate Left", .rotateLeft, key: String(UnicodeScalar(NSLeftArrowFunctionKey)!), in: device)
        command("Rotate Right", .rotateRight, key: String(UnicodeScalar(NSRightArrowFunctionKey)!), in: device)
        device.addItem(.separator())
        command("Home", .home, key: "h", modifiers: [.command, .shift], in: device)
        command("Lock", .lock, key: "l", in: device)
        command("Shake", .shake, key: "z", modifiers: [.command, .control], in: device)
        device.addItem(.separator())
        command("Shut Down", .shutdown, in: device)

        let io = menu("I/O", in: main)
        command("Toggle Software Keyboard", .keyboard, key: "k", in: io)
        command("Open URL…", .openURL, key: "u", modifiers: [.command, .shift], in: io)
        let features = menu("Features", in: main)
        command("Toggle Appearance", .appearance, key: "a", modifiers: [.command, .shift], in: features)

        let window = menu("Window", in: main)
        item("Minimize", action: #selector(minimizeWindow(_:)), key: "m", in: window)
        item("Zoom", action: #selector(zoomWindow(_:)), in: window)
        item("Enter Full Screen", action: #selector(fullScreen(_:)), key: "f", modifiers: [.command, .control], in: window)
        command("Show Device Bezels", .bezel, in: window)
        command("Stay On Top", .stayOnTop, in: window)
        window.addItem(.separator())
        command("Fit Screen", .fit, key: "4", in: window)
        window.addItem(.separator())
        item("Bring All to Front", action: #selector(NSApplication.arrangeInFront(_:)), in: window, target: NSApp)
        NSApp.mainMenu = main
        NSApp.windowsMenu = window
    }

    private func menu(_ title: String, in parent: NSMenu) -> NSMenu {
        let item = NSMenuItem(title: title, action: nil, keyEquivalent: "")
        let menu = NSMenu(title: title)
        item.submenu = menu
        parent.addItem(item)
        return menu
    }

    @discardableResult
    private func item(_ title: String, action: Selector, key: String = "",
                      modifiers: NSEvent.ModifierFlags = .command, in menu: NSMenu,
                      target: AnyObject? = nil) -> NSMenuItem {
        let item = NSMenuItem(title: title, action: action, keyEquivalent: key)
        item.target = target ?? self
        item.keyEquivalentModifierMask = modifiers
        menu.addItem(item)
        return item
    }

    private func command(_ title: String, _ command: DeviceCommand, key: String = "",
                         modifiers: NSEvent.ModifierFlags = .command, in menu: NSMenu) {
        item(title, action: #selector(performCommand(_:)), key: key, modifiers: modifiers, in: menu)
            .representedObject = command.rawValue
    }

    func validateMenuItem(_ item: NSMenuItem) -> Bool {
        if item.action == #selector(showBrowser(_:)) { return true }
        if item.action == #selector(fullScreen(_:)) {
            item.title = NSApp.keyWindow?.styleMask.contains(.fullScreen) == true ? "Exit Full Screen" : "Enter Full Screen"
        }
        guard let value = item.representedObject as? String, let command = DeviceCommand(rawValue: value) else {
            return NSApp.keyWindow != nil
        }
        guard let session = activeSession else { item.state = .off; return false }
        switch command {
        case .screenshot, .rotateLeft, .rotateRight, .importFiles, .openURL: return session.canPerformToolOperation
        case .shutdown: return session.canPerformToolOperation && session.toolbarState.recording == nil
        case .recording:
            item.title = session.toolbarState.recording == nil ? "Record Video…" : "Stop Recording"
            return session.toolbarState.recording.map { !$0.isStopping } ?? session.canPerformToolOperation
        case .bezel: item.state = session.showsDeviceBezel ? .on : .off
        case .stayOnTop: item.state = session.staysOnTop ? .on : .off
        case .fit: return session.canPerformCommands && session.window?.inLiveResize == false
            && session.window?.styleMask.contains(.fullScreen) == false
        default: break
        }
        return session.canPerformCommands
    }

    @objc private func showBrowser(_ sender: Any?) { application?.showDeviceBrowser() }
    @objc private func closeWindow(_ sender: Any?) { NSApp.keyWindow?.performClose(sender) }
    @objc private func minimizeWindow(_ sender: Any?) { NSApp.keyWindow?.performMiniaturize(sender) }
    @objc private func zoomWindow(_ sender: Any?) { NSApp.keyWindow?.performZoom(sender) }
    @objc private func fullScreen(_ sender: Any?) { NSApp.keyWindow?.toggleFullScreen(sender) }
    @objc private func performCommand(_ sender: NSMenuItem) {
        if let raw = sender.representedObject as? String, let command = DeviceCommand(rawValue: raw) {
            activeSession?.perform(command)
        }
    }
}
